### PR TITLE
read into gcp

### PR DESCRIPTION
**Pull request by benny-conn**
Changes:

- **Added** a new type called `FileHeaderReader` that wraps a reader and pre-reads the first 512 bytes so that we can detect the media type of a response without fulling downloading it. The reader can then later be read and will still include the first 512 bytes as if the initial reader was never read.
- **Changed** media processing to not lock and instead use readers to stream data directly into GCP

TODO:

- [x] GLTF detection?
- [x] Figure out if ffmpeg will read videos into memory?
- [x] Testing

### CODE REVIEW

**Comment by radazen** [comment]
LGTM! Excited to see how it goes!

File: service/media/media.go [hunk 1]
```diff
@@ -500,28 +497,29 @@ outer:
...
 	}
 
-	if util.Contains([]string{"development", "sandbox-backend", "production"}, strings.ToLower(viper.GetString("ENV"))) {
-		downloadLock.Lock()
-		defer downloadLock.Unlock()
-	}
+	// if util.Contains([]string{"development", "sandbox-backend", "production"}, strings.ToLower(viper.GetString("ENV"))) {
+	// 	downloadLock.Lock()
+	// 	defer downloadLock.Unlock()
+	// }
```
Comment (Robinnnnn) [inline service/media/media.go:503]: meant to keep this?

File: service/media/media.go [hunk 2]
```diff
@@ -500,28 +497,29 @@ outer:
...
 	if mediaType == persist.MediaTypeUnknown || mediaType == persist.MediaTypeInvalid || mediaType == persist.MediaTypeSyncing {
-		mediaType, contentType = persist.SniffMediaType(bs)
-		if mediaType == persist.MediaTypeUnknown || mediaType == persist.MediaTypeInvalid || mediaType == persist.MediaTypeSyncing {
-			mediaType, contentType = GuessMediaType(bs)
-			logger.For(pCtx).Infof("guessed media type for %s: %s", truncateString(mediaURL, 50), mediaType)
-		} else {
-			logger.For(pCtx).Infof("sniffed media type for %s: %s", truncateString(mediaURL, 50), mediaType)
-		}
+		mediaType, contentType = persist.SniffMediaType(reader.Headers())
+		// if mediaType == persist.MediaTypeUnknown || mediaType == persist.MediaTypeInvalid || mediaType == persist.MediaTypeSyncing {
```
Comment (Robinnnnn) [inline service/media/media.go:517]: same comment

File: service/rpc/rpc.go [hunk 3]
```diff
@@ -325,13 +325,113 @@ func GetDataFromURI(ctx context.Context, turi persist.TokenURI, ipfsClient *shel
...
-		return removeBOM(newImage.Bytes()), nil
+		return util.RemoveBOM(newImage.Bytes()), nil
 	default:
 		return []byte(turi), nil
 	}
 
 }
 
+// GetDataFromURIAsReader calls URI and returns the data as an unread reader with the headers pre-read
+func GetDataFromURIAsReader(ctx context.Context, turi persist.TokenURI, ipfsClient *shell.Shell, arweaveClient *goar.Client) (util.FileHeaderReader, error) {
```
Comment (Robinnnnn) [inline service/rpc/rpc.go:336]: easy way to test this?

**Review by jarrel-b — APPROVED** [review]

File: _encrypted_deploy [hunk 4]
```diff
@@ -1,10 +1,11 @@
...
... 3mkMra789sdH9BI vdhk2yxixHCCexp VIFoQYyt4jSq1p3 Uvfla1L6AvPb3c1 RDlAxDGg7okmfYK 2Q0MjkgRBUNZVUj w3RoPAdESVKsc3D 1yColKDMNatybQt 8xzPfxoqJWDtJYa pTsr950tM7iKaYM tpMrwwpDzLgZ21r YvSf1js8vNqa8hm cv5ouajWsCmW2RY WwGhnmcT0H1OPfa 6pjHPAGDJ37opty kH20tm0Q7MdnXvw XBot0YCeewDVHl7 Aa72ozX4iwKpyBQ AwlOedTmY7Vs6mA wLoYyqoYhtWmzCK UBfA0zoAS6UoOFd y8r9zLDYU1b22vv XQ0uHt214Aw5iH7 RGXqptkFq1fDSnD JlGpes8VJppwD0v hC2NAvhhu30wM6T pw7t9M8Yx59TIbY hxTyw1vlgN5TC9o LZn0CbSnECA2jNt 27KHzpzQpd0LFyO cm5NeLMxEf8zvJI y57iPzaM6r1C200 zyqsEZCAapIQ3Pb e93Ltjha1KgsB3i rXEXQpToJZViXYN YD5pcwuNvcsZIJO aBM48gGZfyGPxMH ArgelDLxEZSEcNb DPMJ7aXfxiCtU65 izOl3JZyhg1wZp5 sOwZaXFyc4S1Qt1 Urt45lmnFwyPuSN VtAvvI2cblo0aLy KvDzfdKuJj3kLhq GZ0wLtcR1Jnvibf 5meuMeFUZEkRN4x xQWuyVxXVJSqpcj Edw69uNHvqEqCh5 qJ0QgkBLaeRkbXA Ju9AeNKVCm1yNwC zDbEuuSq1oVp7Tg pDnDo7ShCW3Sqgx wmQoNhr2sIxmyM5 z9tvqA3aJX6bFd8 FlqIb6r9aBH4hCZ YpTImcCunDxwvmW xCJ87U6xXeYnXC5 tavU43d4dpznQw2 kSgiLgdK1uzJqkm Am1O5k7W9EUxSYV wESUM2uvgZYiErk VZgW6Wl4AWkmaN3 n8we679zfA4Awgc wHpEvRxEoniiEkK qheNpUDKVKOP5V2 rDKBUkQH5LjQ4Pt LQYHdI6lFm1EGLB N0uvFxPAiHBg5DL 23SnHfv674j5cyp cBn9r0gXUQzOdBi wORc9oFrO4jhrpE 88KMhOQl8OlwMjt 66khj3IqE2hr5NY oSRuBXMt5t922JE xWLo7NGGd0eLVNH
```
[the remark this answers is not in the document]
Comment (benny-conn) [reply _encrypted_deploy]: I merged all of the changes into one new deploy 👍🏻

**Pull request merged**